### PR TITLE
VNDA Integration

### PR DESCRIPTION
**Pull request by loliveirawebdev**
added VNDA product loaders and added the capacity to send sort options in listing pages (as optional param for this moment)

### CODE REVIEW

File: commerce/vnda/constants.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { SortOption } from "../types.ts";
+
+export const BASE_URL_PROD = "https://api.vnda.com.br/api/v2/";
```
Comment (tlgimenes) [inline commerce/vnda/constants.ts:3]: No need for these to be on a separate file. Clients could have all of those

File: commerce/vnda/constants.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+export const BASE_URL_SANDBOX = "https://api.sandbox.vnda.com.br/api/v2/";
+export const DOMAIN_HEADER = "X-Shop-Host";
+export const VNDA_SORT_OPTIONS: SortOption[] = [
```
Comment (tlgimenes) [inline commerce/vnda/constants.ts:6]: nice

File: commerce/vnda/transform.ts [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+export const toProduct = (
+  product: ProductGetResultVNDA | ProductVariationVNDA,
+  options: ProductOptions,
+  level = 0,
+  parentUrlData?: { slug: string; productId: string },
+): Product => {
+  const getVariants = () => {
+    if (level !== 0) return [];
+
+    const _product = product as ProductGetResultVNDA;
```
Comment (tlgimenes) [inline commerce/vnda/transform.ts:57]: can we not use casting? Maybe a typeguard is better in here. Also, these functions (getVariants, getSku, getProperties) could be made pure

File: commerce/vnda/transform.ts [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+      lowPrice: product.sale_price,
+      offerCount: 1,
+      offers: [toOffer(product)],
+    },
+  };
+};
+
+const toOffer = (product: ProductBaseVNDA): Offer => {
+  const numberInstallments = product.installments as number[];
+  const sumNumbers = () => numberInstallments.reduce((acc, i) => (acc + i), 0);
```
Comment (tlgimenes) [inline commerce/vnda/transform.ts:125]: prefer pure functions

File: commerce/vnda/transform.ts [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+  return maybeProduct as ProductVariationVNDA;
+};
+
+/**
+ * Não é possível pegar um produto pelo sku, então para que tudo
+ * funcione normalmente, essa função simula o comportamento padrão
+ * de outros ecommerces simplesmente copiando os dados da variação
+ * para o nivel do produto pai.
+ */
+export const useVariant = (
```
Comment (tlgimenes) [inline commerce/vnda/transform.ts:219]: I guess you did it right

In VTEX/Shopify loaders we do the same. The SKU maps to the product and the VTEX's product maps to ProductGroup. 

in here, vnda's sku should map to a deco product

File: commerce/vnda/types.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
+// deno-lint-ignore-file no-explicit-any ban-types
```
Comment (tlgimenes) [inline commerce/vnda/types.ts:1]: use unknown instead of any

File: commerce/vnda/utils/queryBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+// deno-lint-ignore no-explicit-any
+export const paramsToQueryString = (params: Record<string, any>): string => {
```
Comment (tlgimenes) [inline commerce/vnda/utils/queryBuilder.ts:2]: why not use `URLSearchParams` ?

File: commerce/vnda/utils/request.proxy.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  return (params: ConfigVNDA) => {
+    const { domain, authToken, useSandbox } = params;
+    const baseUrl = useSandbox ? BASE_URL_SANDBOX : BASE_URL_PROD;
+
+    // deno-lint-ignore no-explicit-any
+    const fetcher: Fetcher<any> = (endpoint, method, data) => {
+      return fetchAPI(new URL(endpoint, baseUrl).href, {
+        body: data ? JSON.stringify(data) : undefined,
+        method: method || "GET",
+        headers: {
```
Comment (tlgimenes) [inline commerce/vnda/utils/request.proxy.ts:24]: move this to a constant on client.ts

File: functions/vndaProductListingPage.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  const nextPage = new URLSearchParams(url.searchParams);
+  nextPage.set("page", (page + 1).toString());
+
+  const previousPage = new URLSearchParams(url.searchParams);
+  previousPage.set("page", (page - 1).toString());
+
+  return {
+    data: {
+      "@type": "ProductListingPage",
+      // TODO: Find out what's the right breadcrumb on vnda
```
Comment (tlgimenes) [inline functions/vndaProductListingPage.ts:87]: nice, sometimes they don't have any

File: commerce/vnda/client.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  const searchProduct = (params: ProductSearchParams) => {
+    const { type_tags, ...knownParams } = params;
+    const typeTagsEntries = type_tags?.map((tag) => [tag.key, tag.value]) || [];
+
+    const qs = paramsToQueryString({
+      ...knownParams,
+      ...Object.fromEntries(typeTagsEntries),
+    });
+
+    const endpoint = `products/search?show_only_available=true&${qs}`;
```
Comment (lucis) [inline commerce/vnda/client.ts:59]: WDYT about `showOnlyAvailable` being passed as an option (with a default of `true`)?

File: functions/vndaProductDetailsPage.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
+import type { LiveState } from "$live/types.ts";
+import type { LoaderFunction } from "$live/types.ts";
+import { ConfigVNDA } from "../commerce/vnda/types.ts";
+import { createClient } from "../commerce/vnda/client.ts";
+import { toProduct, useVariant } from "../commerce/vnda/transform.ts";
+import type { ProductDetailsPage } from "../commerce/types.ts";
+
+/**
+ * @title VTEX Product Page Loader
+ * @description Works on routes of type /:slug/p
```
Comment (lucis) [inline functions/vndaProductDetailsPage.ts:10]: Is this right that the page schema is `/:slug/p`?

From what I could see, the page would be something like `/product?id=123&skuId=112`, but shouldn't it be more like VNDA's current schema (https://template1.vnda.dev/produto/produto-test-21)?

Usually, having the product's slug in URL is a good SEO pattern

File: commerce/vnda/client.ts (hunk 10 above)
Comment (lucis) [reply commerce/vnda/client.ts:59]: It can be done later, no problems

Comment (loliveirawebdev) [reply commerce/vnda/client.ts:59]: lol
i put this when i was trying to fix the "type_tag" issue then forgot to remove

File: functions/vndaProductDetailsPage.ts (hunk 11 above)
Comment (loliveirawebdev) [reply functions/vndaProductDetailsPage.ts:10]: actually the schema is: [host]/[slug]/p?[id]&[sku]

the VNDA schema concatenates "slug" + "id" as page identifier (product-test-21, where 21 is the product id, for example)... do you think that its better than the schame that i've implemented?

Comment (lucis) [reply functions/vndaProductDetailsPage.ts:10]: Ok, 

So the corresponding URL would be `/product-test?id=21`?

I'm not sure about the implications of it... WDYT @tlgimenes?

File: commerce/vnda/client.ts [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+import { fetchAPI } from "../../utils/fetchAPI.ts";
+import { paramsToQueryString } from "./utils/queryBuilder.ts";
+
+import {
+  ConfigVNDA,
+  ProductGetParams,
+  ProductGetResultVNDA,
+  ProductSearchParams,
+  ProductSearchResultVNDA,
+} from "./types.ts";
```
Comment (mcandeia) [inline commerce/vnda/client.ts:15]: nit: this is a bit unusual to me, having imports as not the first thing of the file, but I'm ok if the other maintains are ok too cc @lucis

File: commerce/vnda/client.ts [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+  const baseUrl = useSandbox ? BASE_URL_SANDBOX : BASE_URL_PROD;
+
+  const fetcher = <T>(
+    endpoint: string,
+    method?: string,
+    data?: Record<string, unknown>,
+  ) => {
+    return fetchAPI<T>(new URL(endpoint, baseUrl).href, {
+      body: data ? JSON.stringify(data) : undefined,
+      method: method || "GET",
```
Comment (mcandeia) [inline commerce/vnda/client.ts:36]:
```suggestion
      method: method ?? "GET",
```

File: commerce/vnda/client.ts [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+  };
+
+  const getProduct = (params: ProductGetParams) => {
+    const endpoint = `products/${params.id}`;
+    return fetcher<ProductGetResultVNDA>(endpoint);
+  };
+
+  const searchProduct = (params: ProductSearchParams) => {
+    const { type_tags, ...knownParams } = params;
+    const typeTagsEntries = type_tags?.map((tag) => [tag.key, tag.value]) || [];
```
Comment (mcandeia) [inline commerce/vnda/client.ts:52]:
```suggestion
    const typeTagsEntries = type_tags?.map((tag) => [tag.key, tag.value]) ?? [];
```

File: commerce/vnda/types.ts [hunk 15]
```diff
@@ -0,0 +1,101 @@
+export interface ConfigVNDA {
+  /**
+   * @description Your VNDA domain name. For example, https://mystore.vnda.com.br
+   */
+  domain: string;
+
+  /**
+   * @description The token generated from admin panel. Read here: https://developers.vnda.com.br/docs/chave-de-acesso-e-requisicoes
+   */
+  authToken: string;
```
[the remark this answers is not in the document]
Comment (lucis) [reply commerce/vnda/types.ts:10]: @mcandeia I think so.

There's an undocumented API used by https://template1.vnda.dev/, but I don't know if it provides all data needed here.

(I think it's a VTEX particularity that Catalog APIs are public. Shopify, for example, requires an Access Token – that can be public though, IIRC)

**Comment by lucis** [comment]
@loliveirawebdev 
image

https://deco-sites-template2.deno.dev/produto-sem-atributo/p?id=564&skuId=teste&pageId=6667

PDP breaks when there's no product. I thought I saw you showing an empty page... Can you take a look if that regressed?

File: commerce/vnda/client.ts (hunk 12 above)
Comment (loliveirawebdev) [reply commerce/vnda/client.ts:15]: actually, this is an anti-pattern, right?
i just forgot to re-order when i was moving these constants from another file

File: commerce/vnda/types.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (loliveirawebdev) [reply commerce/vnda/types.ts:10]: yes... this is strictly necessary :/ i will add your suggestion

**Comment by loliveirawebdev** [comment]
@lucis fixed it
@mcandeia great! i'll create an unit test for it

**Comment by lucis** [comment]
LGTM, @loliveirawebdev!

I'll merge it right now to be able to work on top of it, but feel free to complement it if anything comes up

**Pull request merged**